core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -821,9 +858,119 @@ object Parser {
 
   /** Represents where a failure occurred and all the expectations that were broken
     */
-  final case class Error(failedAtOffset: Int, expected: NonEmptyList[Expectation]) {
+  final class ErrorWithInput(
+      val input: String,
+      override val failedAtOffset: Int,
+      override val expected: NonEmptyList[Expectation]
+  ) extends Error(failedAtOffset, expected) {}
+
+  class Error(val failedAtOffset: Int, val expected: NonEmptyList[Expectation])
```

Comment:
if we go this route, I'd rather a single class:

```scala
class Error(val input: Option[String], val failedAtOffset: Int, val expected: NonEmptyList[Expectation]) {
  def this(failedAtOffset: Int, expected: NonEmptyList[Expectation]) =
    this(None, failedAtOffset, expected)
}
```
then we can also include the `input: Option[String]` in the `def equals`.